### PR TITLE
Merak-Topo: give namespace for each topo, rewrite log and error handling

### PR DESCRIPTION
**Pull request by zzengcs**
In this PR, the major changes contain:
1. merak-topo can create topology under "merak-topoid" namespace
2. namespace parameter is set in a single entry
3. add initial compute node and host node data in the DB
4. refine log message with multiple categories as Info, Debug, Warn, Error.
5. refine the error handling

This PR addresses all the issues covered in the PR #173 
This PR passes the end2end test in Kind env and K8s cluster.

### CODE REVIEW

File: services/merak-topo/handler/handler.go [hunk 1]
```diff
@@ -15,70 +15,89 @@ package handler
...
 
-	err_create, topo := Create_multiple_layers_vswitches(int(aca_num), int(rack_num), int(aca_per_rack), int(ports_per_vswitch), data_plane_cidr)
+	topo, err_create := Create_multiple_layers_vswitches(int(aca_num), int(rack_num), int(aca_per_rack), int(ports_per_vswitch), data_plane_cidr)
 	if err_create != nil {
-		return fmt.Errorf("create multiple layers vswitches error %s", err_create)
+		utils.Logger.Error("request DEPLOY", "multiple layers vswitches", err_create.Error())
+		returnMessage.ReturnMessage = "Can not create multiple layers vswitches"
+		returnMessage.ReturnCode = pb_common.ReturnCode_FAILED
+		err_flag = 1
+
```
Comment (cj-chung) [inline services/merak-topo/handler/handler.go:61]: @zzengcs Is that mean even we have error from `Create_multiple_layers_vswitches`, the logic here still go on for the rest of process? Is that safe?

File: services/merak-topo/handler/handler.go [hunk 2]
```diff
@@ -15,70 +15,89 @@ package handler
...
-		return fmt.Errorf("CREATE: failed to list k8s host nodes info %s", err1)
+	if err2 != nil {
+		utils.Logger.Error("request DEPLOY", "check nodes in k8s cluster", err2.Error())
+		returnMessage.ReturnMessage = "CREATE: can not list k8s host nodes info"
+		returnMessage.ReturnCode = pb_common.ReturnCode_FAILED
+		err_flag = 1
 	}
 
 	for _, s := range k8s_nodes.Items {
 		var hnode database.HostNode
```
Comment (cj-chung) [inline services/merak-topo/handler/handler.go:94]: @zzengcs if `err_flag=1`, that mean k8s create nodes failed and `k8s_nodes` will be empty. Then the next `for` loop will get nil pointer for `k8s_nodes.Items`.

File: services/merak-topo/handler/handler.go [hunk 3]
```diff
@@ -15,70 +15,89 @@ package handler
...
 
-		node_yaml, err2 := k8client.CoreV1().Nodes().Get(Ctx, s.Name, metav1.GetOptions{})
-		if err2 != nil {
-			return fmt.Errorf("CREATE: get k8s host node info from k8s cluster error %s", err2)
+		node_yaml, err3 := k8client.CoreV1().Nodes().Get(Ctx, s.Name, metav1.GetOptions{})
+		if err3 != nil {
+			utils.Logger.Error("can't get host node info in k8s cluster", s.Name, err3.Error())
+			returnMessage.ReturnMessage = "DEPLOY: can not get k8s host node info from k8s cluster"
+			returnMessage.ReturnCode = pb_common.ReturnCode_FAILED
+			err_flag = 1
```
Comment (cj-chung) [inline services/merak-topo/handler/handler.go:100]: @zzengcs same here. If `err_flag=1`, `node_yaml` will be nil, then the following 2nd `for` loop will complain about `node_yaml` is nil.

**Review by cj-chung — CHANGES_REQUESTED** [review]

File: services/merak-topo/handler/handler.go [hunk 4]
```diff
@@ -224,31 +200,54 @@ func Create(k8client *kubernetes.Clientset, topo_id string, aca_num uint32, rack
...
+
+	elaps2 := time.Since(start1)
+
+	if err_flag == 1 {
+		err_return = errs
+		utils.Logger.Info("request DEPLOY", "Fails: check merak-topo logs", elaps2)
+	} else {
+		err_return = nil
+		utils.Logger.Info("request DEPLOY", "Complete: Return compute nodes and k8s cluster information to Scenario Manager (in second)", elaps2)
+	}
```
Comment (cj-chung) [inline services/merak-topo/handler/handler.go:219]: @zzengcs Is that really need to use `err_flag` in the end of function to identity all errors from above?

File: services/merak-topo/handler/handler.go [hunk 5]
```diff
@@ -224,31 +200,54 @@ func Create(k8client *kubernetes.Clientset, topo_id string, aca_num uint32, rack
...
 	}
 
 	k8s_nodes, err1 := k8client.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
 	if err1 != nil {
-		return fmt.Errorf("UpdateComputenodeInfo: failed to list k8s host nodes info %s", err1)
+		utils.Logger.Error("can't check k8s cluster ", "corev1", err1.Error())
+		err_flag = 1
 	}
 
 	for _, s := range k8s_nodes.Items {
```
Comment (cj-chung) [inline services/merak-topo/handler/handler.go:244]: @zzengcs same here. If `err_flag=`, `k8s_nodes` will be nil or empty and the following `for` loop will complain about it.

File: services/merak-topo/handler/handler.go [hunk 6]
```diff
@@ -287,18 +286,17 @@ func UpdateComputenodeInfo(k8client *kubernetes.Clientset, topo_id string) error
...
 
 			var cnode database.ComputeNode
 
-			res, err := k8client.CoreV1().Pods("default").Get(Ctx, node.Name, metav1.GetOptions{})
+			res, err := k8client.CoreV1().Pods(namespace).Get(Ctx, node.Name, metav1.GetOptions{})
 
 			if err != nil {
-				log.Printf("updatecomputenode: get pod info from k8s error %s", err)
+				utils.Logger.Error("can't get pod info from k8s", node.Name, err.Error(), "namespace", namespace)
+				err_flag = 1
```
Comment (cj-chung) [inline services/merak-topo/handler/handler.go:299]: @zzengcs same here...

File: services/merak-topo/handler/handler.go [hunk 7]
```diff
@@ -346,33 +345,45 @@ func UpdateComputenodeInfo(k8client *kubernetes.Clientset, topo_id string) error
...
-		return fmt.Errorf("update compute node info: query topology_id error %s", err)
+		utils.Logger.Warn("request CHECK", topoPrefix, err.Error())
 	}
 
 	k8s_nodes, err1 := k8client.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
 
 	if err1 != nil {
-		return fmt.Errorf("INFO: list k8s host nodes info error %s", err1)
+		utils.Logger.Error("k8s cluster no response ", "corev1", err1.Error())
+		err_flag = 1
```
Comment (cj-chung) [inline services/merak-topo/handler/handler.go:375]: @zzengcs Same here!!

File: services/merak-topo/handler/handler.go [hunk 8]
```diff
@@ -429,39 +436,44 @@ func Info(k8client *kubernetes.Clientset, topo_id string, returnMessage *pb.Retu
...
-		return fmt.Errorf("query topology_id error %s", err_db)
+		utils.Logger.Warn("request DELETE", "can't query topology data from DB", topoPrefix, err_db.Error())
 	}
 
 	k8s_nodes, err1 := k8client.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
 
 	if err1 != nil {
-		return fmt.Errorf("INFO: list k8s host nodes info error %s", err1)
+		utils.Logger.Error("request DELETE", "k8s cluster no response", err1.Error())
+		return err1
```
Comment (cj-chung) [inline services/merak-topo/handler/handler.go:464]: @zzengcs Same here!!

File: services/merak-topo/handler/topo_deploy.go [hunk 9]
```diff
@@ -28,72 +26,60 @@ import (
...
+	rc := NewTopologyClass(name, links, namespace)
+
+	var err_return error
 
 	_, err := client.Resource(topologyClassGVR).Namespace(namespace).Create(Ctx, rc, metav1.CreateOptions{})
 
 	if err != nil {
-		return fmt.Errorf("failed to create topologyClass %s", err)
+		utils.Logger.Error("can't create topologyClass", "create topology class error", err.Error(), "namespace", namespace, "vnode", name)
+		err_return = err
```
Comment (cj-chung) [inline services/merak-topo/handler/topo_deploy.go:58]: @zzengcs just `return err_return` should be enough.

File: services/merak-topo/handler/topo_deploy.go [hunk 10]
```diff
@@ -28,72 +26,60 @@ import (
...
 	err := client.Resource(topologyClassGVR).Namespace(namespace).Delete(Ctx, name, metav1.DeleteOptions{})
 
-	log.Printf("Delete TopologyClass %s", name)
-
 	if err != nil {
-		return fmt.Errorf("failed to create topologyClass %s", err)
+		utils.Logger.Error("can't delete topologyClass", "topology class deletion error", err.Error(), "namespace", namespace, "vnode", name)
+		err_return = err
+	} else {
+		err_return = nil
```
Comment (cj-chung) [inline services/merak-topo/handler/topo_deploy.go:76]: @zzengcs same here, just `return err_return` should be enough.

File: services/merak-topo/grpc/service/service.go [hunk 11]
```diff
@@ -108,10 +117,12 @@ func (s *Server) TopologyHandler(ctx context.Context, in *pb.InternalTopologyInf
 		// if data_plane_cidr == "" || aca_num == 0 || aca_per_rack == 0 || rack_num == 0 || cgw_num == 0 {
```
[the remark this answers is not in the document]
Comment (zzengcs) [reply services/merak-topo/grpc/service/service.go:117]: addressed

File: services/merak-topo/grpc/service/service.go [hunk 12]
```diff
@@ -128,53 +139,64 @@ func (s *Server) TopologyHandler(ctx context.Context, in *pb.InternalTopologyInf
...
-	log.Printf("Received on TopologyHandler %s", in)
-	var returnMessage pb.ReturnTopologyMessage
-	return &returnMessage, nil
 }
+
+// func (s *Server) TestHandler(ctx context.Context, in *pb.InternalTopologyInfo) (*pb.ReturnTopologyMessage, error) {
+// 	log.Printf("Received on TopologyHandler %s", in)
+// 	var returnMessage pb.ReturnTopologyMessage
+// 	return &returnMessage, nil
+// }
```
[the remark this answers is not in the document]
Comment (zzengcs) [reply services/merak-topo/grpc/service/service.go:202]: addressed

**Review by cj-chung — CHANGES_REQUESTED** [review]

File: services/merak-topo/handler/topo_deploy.go [hunk 13]
```diff
@@ -441,88 +432,107 @@ func ovs_config(topo database.TopologyData, node_name string, sdn_ip string, sdn
...
-	// check pod status
-	topo_id := topo.Topology_id
+			err_del := k8client.CoreV1().Pods(namespace).Delete(Ctx, node.Name, metav1.DeleteOptions{})
 
-	err_db := database.SetValue(topo_id, topo)
-	if err_db != nil {
-		return fmt.Errorf("fail to save in db %s", err_db)
+			if err_del != nil {
+				utils.Logger.Error("can't delete topology pod in k8s cluster", "pod name", node.Name, "namespace", namespace, "error msg", err_del.Error())
+				err_flag = 1
```
Comment (cj-chung) [inline services/merak-topo/handler/topo_deploy.go:470]: @zzengcs just `return err_return` here.

File: services/merak-topo/handler/handler.go (hunk 3 above)
Comment (zzengcs) [reply services/merak-topo/handler/handler.go:100]: addressed-- return err when checking k8s cluster worker nodes does not work

File: services/merak-topo/handler/topo_deploy.go [hunk 14]
```diff
@@ -441,88 +432,107 @@ func ovs_config(topo database.TopologyData, node_name string, sdn_ip string, sdn
...
+			}
+
+			err_del_t := DeleteTopologyClasses(dclient, node.Name, namespace)
+			if err_del_t != nil {
+				utils.Logger.Error("can't delete topology class in meshnet", "pod name", node.Name, "namespace", namespace, "error msg", err_del_t.Error())
+				err_flag = 1
```
Comment (cj-chung) [inline services/merak-topo/handler/topo_deploy.go:476]: @zzengcs just `return err_return` here.

File: services/merak-topo/handler/topo_deploy.go [hunk 15]
```diff
@@ -441,88 +432,107 @@ func ovs_config(topo database.TopologyData, node_name string, sdn_ip string, sdn
...
+			}
+
+		}
+	}
+
+	if err_flag == 1 {
+		return err_return
+	} else {
+		return nil
```
Comment (cj-chung) [inline services/merak-topo/handler/topo_deploy.go:485]: @zzengcs just `return nil` here.

File: services/merak-topo/handler/topo_deploy.go [hunk 16]
```diff
@@ -441,88 +432,107 @@ func ovs_config(topo database.TopologyData, node_name string, sdn_ip string, sdn
...
 	}
-	return nil
 }
+
+/*Comment: unused function*/
+
+// func Pod_query(k8client *kubernetes.Clientset, pod *corev1.Pod, cmd []string, namespace string) (string, error) {
+
```
Comment (cj-chung) [inline services/merak-topo/handler/topo_deploy.go:492]: @zzengcs if no need it, just remove it.

File: services/merak-topo/handler/handler.go (hunk 4 above)
Comment (zzengcs) [reply services/merak-topo/handler/handler.go:219]: @cj-chung using err_flag is to give user the readable error info by skipping the details. The detailed error msgs are recorded by log. If don't use it, any suggestions?

File: services/merak-topo/utils/utils.go [hunk 17]
```diff
@@ -40,11 +57,15 @@ func K8sClient() (*kubernetes.Clientset, error) {
...
 func K8sConfig() (*rest.Config, error) {
 
+	var err_return error
+	err_return = nil
+
 	config, err := rest.InClusterConfig()
 	if err != nil {
-		panic(err.Error())
+		Logger.Fatal("can't create k8s config", "k8s config initiate error", err.Error())
+		err_return = err
```
Comment (cj-chung) [inline services/merak-topo/utils/utils.go:66]: @zzengcs just `return nil, err` here

File: services/merak-topo/handler/handler.go (hunk 5 above)
Comment (zzengcs) [reply services/merak-topo/handler/handler.go:244]: addressed-- return err when checking k8s cluster worker nodes does not work

File: services/merak-topo/utils/utils.go [hunk 18]
```diff
@@ -40,11 +57,15 @@ func K8sClient() (*kubernetes.Clientset, error) {
...
 	}
 
-	return config, nil
+	return config, err_return
```
Comment (cj-chung) [inline services/merak-topo/utils/utils.go:69]: @zzengcs just `return config, nil` here.

File: services/merak-topo/utils/utils.go [hunk 19]
```diff
@@ -40,11 +57,15 @@ func K8sClient() (*kubernetes.Clientset, error) {
 
 func K8sConfig() (*rest.Config, error) {
 
+	var err_return error
+	err_return = nil
```
Comment (cj-chung) [inline services/merak-topo/utils/utils.go:61]: @zzengcs you don't need  `err_return` here.

**Review by cj-chung — CHANGES_REQUESTED** [review]

**Comment by cj-chung** [comment]
@zzengcs After address all comments, please rerun the test again. Thanks.

File: services/merak-topo/handler/handler.go (hunk 6 above)
Comment (zzengcs) [reply services/merak-topo/handler/handler.go:299]: addressed-- return err when querying k8s cluster worker nodes does not work

File: services/merak-topo/handler/handler.go (hunk 7 above)
Comment (zzengcs) [reply services/merak-topo/handler/handler.go:375]: Yes, it's the same. All similar log msg now are changed 
1) for all k8s env query/check---- return err when fails
2) for all merak-topo DB query/check -- use error flag to return err at end of function. 

@cj-chung , please give some comments if the design above is not appropriate. Thanks!

File: services/merak-topo/handler/handler.go (hunk 8 above)
Comment (zzengcs) [reply services/merak-topo/handler/handler.go:464]: It returns err when k8s cluster has no response on worker nodes. What's the issue here? Thank! @cj-chung

File: services/merak-topo/handler/topo_deploy.go (hunk 9 above)
Comment (zzengcs) [reply services/merak-topo/handler/topo_deploy.go:58]: addressed

File: services/merak-topo/handler/topo_deploy.go (hunk 10 above)
Comment (zzengcs) [reply services/merak-topo/handler/topo_deploy.go:76]: addressed, tks

File: services/merak-topo/handler/topo_deploy.go (hunk 13 above)
Comment (zzengcs) [reply services/merak-topo/handler/topo_deploy.go:470]: addressed, tks

File: services/merak-topo/handler/topo_deploy.go (hunk 14 above)
Comment (zzengcs) [reply services/merak-topo/handler/topo_deploy.go:476]: addressed, tks

File: services/merak-topo/handler/topo_deploy.go (hunk 15 above)
Comment (zzengcs) [reply services/merak-topo/handler/topo_deploy.go:485]: addressed, tks

File: services/merak-topo/handler/topo_deploy.go (hunk 16 above)
Comment (zzengcs) [reply services/merak-topo/handler/topo_deploy.go:492]: addressed, tks

File: services/merak-topo/utils/utils.go (hunk 17 above)
Comment (zzengcs) [reply services/merak-topo/utils/utils.go:66]: addressed, tks

File: services/merak-topo/utils/utils.go (hunk 18 above)
Comment (zzengcs) [reply services/merak-topo/utils/utils.go:69]: return config, err ?

File: services/merak-topo/utils/utils.go (hunk 19 above)
Comment (zzengcs) [reply services/merak-topo/utils/utils.go:61]: it has been removed when addressing the previous two comments, tks

**Comment by cj-chung** [comment]
@zzengcs please make a PR smaller next time, one PR for one specific feature or bug fix.

**Review by phudtran — CHANGES_REQUESTED** [review]
Some minor style related comments thanks. Since go is strongly typed, we don't need to initialize its string variables. They will be empty strings at declaration. If you need to initialize a struct please consider using a constructor. Also, if you use a string in more than 2 place consider making it a constant.

File: services/merak-topo/database/database.go [hunk 20]
```diff
@@ -50,27 +52,36 @@ func ConnectDatabase() error {
...
+		utils.Logger.Warn("can't save key in DB, please retry", "Warning ", err2.Error(), "key", key)
+
 	}
 
 	return nil
 }
 
 func Get(key string) (string, error) {
+
+	val := "XXX"
```
[the remark this answers is not in the document]
Comment (zzengcs) [reply services/merak-topo/database/database.go]: addressed, tks

File: services/merak-topo/database/database.go [hunk 21]
```diff
@@ -79,119 +90,117 @@ func Del(key string) error {
...
 
 	}
 
-	return nil
+	return entity, nil
 }
 
 func getKeys(prefix string) ([]string, error) {
-	var allkeys []string
+	allkeys := []string{"XXX"}
```
[the remark this answers is not in the document]
Comment (zzengcs) [reply services/merak-topo/database/database.go]: addressed, tks

File: services/merak-topo/database/database.go [hunk 22]
```diff
@@ -79,119 +90,117 @@ func Del(key string) error {
...
+
 	}
 	return nil
 }
 
 func FindHostEntity(id string, prefix string) (HostNode, error) {
 	var entity HostNode
+	entity.Ip = "X.X.X.X"
+	entity.Status = STATUS_NONE
+	entity.Routing_rule = []string{"Initial"}
```
[the remark this answers is not in the document]
Comment (zzengcs) [reply services/merak-topo/database/database.go]: addressed, tks

File: services/merak-topo/database/database.go [hunk 23]
```diff
@@ -79,119 +90,117 @@ func Del(key string) error {
...
-
-	values, err2 := getKeyAndValueMap(keys, prefix)
-	if err2 != nil {
-		return nil, fmt.Errorf("GetAllValuesWithKeyPrefix:get key and value map error %s", err2.Error())
-	}
-	return values, nil
-}
+func FindTopoEntity(id string, prefix string) (TopologyData, error) {
+	var entity TopologyData
+	entity.Topology_id = "XXX"
```
[the remark this answers is not in the document]
Comment (zzengcs) [reply services/merak-topo/database/database.go]: addressed, tks

**Review by cj-chung — APPROVED** [review]
LGTM

File: services/merak-topo/utils/utils.go [hunk 24]
```diff
@@ -14,24 +14,41 @@ Copyright(c) 2022 Futurewei Cloud
...
 const K8SAPIURL string = "https://kubernetes.default:6443"
 
+func Init_logger() {
+	var err error
+
+	Logger, err = logger.NewLogger(logger.DEBUG, "Merak-topo")
+
+	if err != nil {
+		Logger.Fatal("Can not build a new logger", err)
+	} else {
```
[the remark this answers is not in the document]
Comment (phudtran) [reply services/merak-topo/utils/utils.go:36]: Please change this from 
```
if err != nil {
     Logger.Fatal("Can not build a new logger", err)
 } else {
     Logger.Info("Create logger for merak-topo", "Init_logger", "Created")
 }
```

to

```
if err != nil {
     Logger.Fatal("Can not build a new logger", err)
 } 
Logger.Info("Create logger for merak-topo", "Init_logger", "Created")
 
```

**Review by phudtran — APPROVED** [review]
LGTM to me. One minor comment.

**Review by cj-chung — APPROVED** [review]
LGTM

**Pull request merged**